aten/src/ATen/copy_wrapper.py

```diff
@@ -19,90 +16,199 @@
+
+    for dst_type in all_types:
+        if dst_type['Density'] == 'Sparse' or src_type['Density'] == 'Sparse':
+            # skip sparse copies, which are not yet implemented
+            continue
         cuda = ''
+        state = []
         if src_type['Backend'] == 'CUDA':
             cuda = 'Cuda'
-        if env['Backend'] == 'CUDA' or src_type['Backend'] == 'CUDA':
-            state.append('context->thc_state')
+        if dst_type['Backend'] == 'CUDA' or src_type['Backend'] == 'CUDA':
+            state = ['context->thc_state']
 
-        combined = nested_dict({
+        body_env = nested_dict({
             'src_scalar_name': src_type['ScalarName'],
-            'src_id': src_type['TypeID'],
+            'case_id': dst_type['TypeID'],
             'src_tensor': src_type['Tensor'],
+            'dst_tensor': dst_type['Tensor'],
             'cuda': cuda,
             'state': state,
-        }, env)
+        }, dst_type)
 
         copies = []
-        if env['ScalarType'] == src_type['ScalarType']:
-            if env['Backend'] == 'CUDA' and src_type['Backend'] == 'CPU':
-                copies.append(COPY_ASYNC_CPU.substitute(combined))
-            if env['Backend'] == 'CPU' and src_type['Backend'] == 'CUDA':
-                copies.append(COPY_ASYNC_CUDA.substitute(combined))
-        copies.append(COPY.substitute(combined))
-
-        copy_body.append(CASE.substitute(combined, copies=copies))
-    return FUNCTION.substitute(env, copy_body=copy_body)
+        if dst_type['ScalarType'] == src_type['ScalarType']:
+            if dst_type['Backend'] == 'CPU':
+                copies.append(COPY_ASYNC_CUDA.substitute(body_env))
+        copies.append(COPY.substitute(body_env))
+
+        copy_body.append(CASE.substitute(body_env, copies=copies))
 
+    return FUNCTION_FROM.substitute(src_type, copy_body=copy_body)
 
-def create(all_types):
 
+def create(all_types, backend):
     top_env = {
         'copy_includes': [],
         'copy_functions': [],
+        'cuda_includes': [],
     }
-    for dst_type in all_types:
+    if backend == 'CPU':
+        # NB: We are going to iterate over this multiple times; better not use a
+        # generator
+        # NB: Backend is 'CPU' or 'SparseCPU'
+        all_types = list(filter((lambda t: t['DenseBackend'] == 'CPU'), all_types))
+    else:
+        top_env['cuda_includes'].append(CUDA_INCLUDES)
+    for the_type in all_types:
         top_env['copy_includes'].append(
-            '#include "ATen/{}.h"'.format(dst_type['Type']))
+            '#include "ATen/{}.h"'.format(the_type['Type']))
         top_env['copy_includes'].append(
-            '#include "ATen/{}.h"'.format(dst_type['Tensor']))
-        top_env['copy_functions'].append(create_one(dst_type, all_types))
+            '#include "ATen/{}.h"'.format(the_type['Tensor']))
+        # Headers for everything in all_types, but generate code only for your
+        # backend
+        if the_type['DenseBackend'] != backend:
```

Comment:
this and writing to all_types seems like it should be pushed down to create_one_copy (or just have an if before it and an assert in it so you don't have to mess with append)?  You are renaming what all_types means and exiting early out of the for loop, which makes this whole function harder to follow.